flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/threadinfo/ThreadInfoRequestCoordinator.java

```diff
@@ -0,0 +1,188 @@
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/** A coordinator for triggering and collecting thread info stats of running operator subtasks. */
+public class ThreadInfoRequestCoordinator
+        extends TaskStatsRequestCoordinator<List<ThreadInfoSample>, OperatorThreadInfoStats> {
+
+    /**
+     * Creates a new coordinator for the job.
+     *
+     * @param executor Used to execute the futures.
+     * @param requestTimeout Time out after the expected sampling duration. This is added to the
+     *     expected duration of a request, which is determined by the number of samples and the
+     *     delay between each sample.
+     */
+    public ThreadInfoRequestCoordinator(Executor executor, long requestTimeout) {
+        super(executor, requestTimeout);
+    }
+
+    /**
+     * Triggers collection of thread info stats of an operator by combining thread info responses
+     * from given subtasks. A thread info response of a subtask in turn consists of {@code
+     * numSamples}, collected with {@code delayBetweenSamples} milliseconds delay between them.
+     *
+     * @param subtasksWithGateways Execution vertices together with TaskExecutors running them.
+     * @param numSamples Number of thread info samples to collect from each subtask.
+     * @param delayBetweenSamples Delay between consecutive samples (ms).
+     * @param maxStackTraceDepth Maximum depth of the stack traces collected within thread info
+     *     samples.
+     * @return A future of the completed thread info stats.
+     */
+    public CompletableFuture<OperatorThreadInfoStats> triggerThreadInfoRequest(
+            List<Tuple2<AccessExecutionVertex, CompletableFuture<TaskExecutorGateway>>>
+                    subtasksWithGateways,
+            int numSamples,
+            Time delayBetweenSamples,
+            int maxStackTraceDepth) {
+
+        checkNotNull(subtasksWithGateways, "Tasks to sample");
+        checkArgument(subtasksWithGateways.size() > 0, "No tasks to sample");
+        checkArgument(numSamples >= 1, "No number of samples");
+        checkArgument(maxStackTraceDepth >= 0, "Negative maximum stack trace depth");
+
+        // Execution IDs of running tasks
+        List<ExecutionAttemptID> runningSubtasksIds = new ArrayList<>();
+
+        // Check that all tasks are RUNNING before triggering anything. The
+        // triggering can still fail.
+        for (Tuple2<AccessExecutionVertex, CompletableFuture<TaskExecutorGateway>>
+                executionsWithGateway : subtasksWithGateways) {
+            AccessExecution execution = executionsWithGateway.f0.getCurrentExecutionAttempt();
+            if (execution != null && execution.getState() == ExecutionState.RUNNING) {
+                runningSubtasksIds.add(execution.getAttemptId());
+            } else {
+                return FutureUtils.completedExceptionally(
+                        new IllegalStateException(
```

Comment:
Will failing so hard not creating quite a bit of noise?

Reply:
Probabaly... That's what the old backpressure sampler was doing https://github.com/apache/flink/blob/7e4694b8198664300b13b5304c62271b192d4512/flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/StackTraceSampleCoordinator.java#L129

What would be the best practice here in your opinion?

Reply:
Tracing on caller side now. Semantic change: Flame graphs are not aborted for tasks with finished subtasks - I think that this is useful for debugging strugglers.